src/rdkafka_broker.c

```diff
@@ -5725,13 +5744,51 @@ void rd_kafka_connect_any(rd_kafka_t *rk, const char *reason) {
                 return;
         }
 
-        /* First pass:  only match brokers never connected to,
+        /* 90% of times select a learned broker in init state.
+         *
+         * This avoid problems after re-bootstrapping that cause
+         * the bootstrap brokers to be always preferred
+         * given there are learned brokers that caused ALL_BROKERS_DOWN.
+         *
+         * If that happens the learned brokers
+         * that already connected are never selected.
+         *
+         * Additionally we cannot always prefer the learned
+         * brokers as their address could have changed and we need to
+         * connect to the bootstrap brokers to know that.
+         */
+        if (rd_jitter(0, 9) > 0) {
+
+                /* First pass:  only match learned brokers never connected to
+                 *              in state INIT, to try to exhaust
+                 *              the available brokers so that an
+                 *              ERR_ALL_BROKERS_DOWN error can be raised. */
+                rkb = rd_kafka_broker_random(
+                    rk, RD_KAFKA_BROKER_STATE_INIT,
+                    rd_kafka_broker_filter_learned_never_connected, NULL);
+
+                if (!rkb)
+                        /* Second pass:  only match learned brokers
+                         *               in state INIT. */
+                        rkb = rd_kafka_broker_random(
+                            rk, RD_KAFKA_BROKER_STATE_INIT,
+                            rd_kafka_broker_filter_learned, NULL);
+#if ENABLE_DEVEL == 1
+                if (rkb)
```

Comment:
Why is this check needed (even in dev only builds) when we are sure that both rd_kafka_broker_filter_learned_never_connected and rd_kafka_broker_filter_learned can only yield learned brokers?